gir/pkgconfig/pkgconfig.go

```diff
@@ -1,105 +1,234 @@
+// Copyright (C) 2021, 2023  diamondburned
+// Copyright (C) 2023-2024  Luke T. Shumaker
+
 // Package pkgconfig provides a wrapper around the pkg-config binary.
 package pkgconfig
 
 import (
 	"errors"
 	"fmt"
-	"io/fs"
-	"os"
 	"os/exec"
 	"path/filepath"
 	"strings"
 	"sync"
 )
 
-var (
-	path     string
-	pathErr  error
-	pathOnce sync.Once
-)
+// A ValueSource is a function that takes a list of package names and
+// returns a map of package-name to some value.
+type ValueSource = func(pkgs ...string) (map[string]string, error)
 
-func ensure() error {
-	pathOnce.Do(func() {
-		path, pathErr = exec.LookPath("pkg-config")
-		if pathErr != nil {
-			pathErr = fmt.Errorf("failed to find pkg-config: %w", pathErr)
+var pkgConfigVer = sync.OnceValues[string, error](func() (string, error) {
+	var stdout strings.Builder
+	cmd := exec.Command("pkg-config", "--version")
+	cmd.Stdout = &stdout
+	if err := cmd.Run(); err != nil {
+		var exitErr *exec.ExitError
+		if !errors.As(err, &exitErr) {
+			return "", fmt.Errorf("pkg-config failed: %w", err)
 		}
-	})
-	return pathErr
-}
+		return "", fmt.Errorf("pkg-config failed with status %d:\n%s",
+			exitErr.ExitCode(), exitErr.Stderr)
+	}
+	return strings.TrimRight(stdout.String(), "\n"), nil
+})
+
+// Values returns a map of package-name to variable-value of the given
+// variable name for each requested package.  It is not an error for a
+// variable to be unset or empty; ret[pkgname] is an empty string if
+// that package does not have the variable set.
+func Values(varname string, pkgs ...string) (map[string]string, error) {
+	if len(pkgs) == 0 {
+		return nil, nil
+	}
 
-// IncludeDir finds the include directories for the given packages.
-func IncludeDirs(pkgs ...string) ([]string, error) {
-	if err := ensure(); err != nil {
+	// There are 3 pkg-config implementations that we should work with:
+	//
+	//  1. FDO `pkg-config`
```

Comment:
Where are these actually used? This repository should only be concerned with whatever is on Nixpkgs (`pkg-config`) imo, so these shouldn't be needed. Detecting version is also not reliable, since nothing prevents these tools from being updated to the next major/minor version.

Reply:
Selfishly, I would like it to work on my Parabola (like Arch Linux) laptop without Nix. Arch hasn't shipped FDO pkg-config since 2018-05-31.

As for reliability: FDO pkg-config hasn't had an update in 7 years; jumping from 0.29 to 1.9.x or 2.x at this point seems exceedingly unlikely; and pkgconf aren't going to conflict with their own version numbers. Though pkgconf could update to 3.x--I hope to convince them to properly support querying multiple packages by then :)

Reply:
No worries. I'll drop it from the PR (might be a few days though, I'm traveling).

But, full disclosure: My main interest in the project is improving the generator as a tool for arbitrary GIR libraries, not so much with the specific GTK bindings.